### PR TITLE
MvnProtocolHandlerService: Implement lastModfified()

### LINKED ISSUE #982: Many 'Server returned lastModified <= 0' warnings when using the 'mvn:' syntax for update site URLs
When using the `mvn:` syntax for update site URLs, many `Server returned lastModified <= 0 for ...` are produced.

E.g.:
image

The reason is that `MvnProtocolHandlerService` does not override `java.net.URLConnection.getLastModified()` and always returns `0`.

I suppose we could simply return `lastModified()` of the resolved `File` referenced within `artifactResult`?

### PR DESCRIPTION
**Pull request by sratz**
Determine last-modified timestamp based on the referenced file.

This silences warnings such as
  ...
  [INFO] Adding repository mvn:org.eclipse.jetty:jetty-p2:11.0.12:zip:p2site
  [WARNING] [a7f3f9cb-793c-4f52-8e8e-b74719152388][extension>org.eclipse.tycho:tycho-maven-plugin:3.0.0] Server returned lastModified <= 0 for mvn:org.eclipse.jetty:jetty-p2:11.0.12:zip:p2site/content.xml
  ...

Fixes #982.

### CODE REVIEW

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MvnProtocolHandlerService.java [hunk 1]
```diff
@@ -129,6 +129,19 @@ public InputStream getInputStream() throws IOException {
       String urlSpec = "jar:" + location.toURI() + "!" + subPath;
       return new URL(urlSpec).openStream();
     }
+
+    @Override
+    public long getLastModified() {
+      try {
+        connect();
+      } catch (IOException e) {
+        return 0;
```
Comment (laeubi) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MvnProtocolHandlerService.java:138]: I think -1 has to be returned for "cant determine last modified"

**Comment by laeubi** [comment]
@sratz The log output don't seem to be related to m2e even though an implementation of lastModified is fine it might require an equal one for Tycho as well.

**Comment by sratz** [comment]
The warning is written by p2:
https://git.eclipse.org/r/plugins/gitiles/equinox/rt.equinox.p2/+/master/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/internal/p2/repository/CacheManager.java#112

**Comment by laeubi** [comment]
But m2e do not use P2 ;-)

**Comment by sratz** [comment]
> But m2e do not use P2 ;-)

No, but P2 uses m2e-core. Indirectly, that is, if you use the `mvn:` URI scheme in a `.target` definition

**Comment by laeubi** [comment]
> > But m2e do not use P2 ;-)
> 
> No, but P2 uses m2e-core. Indirectly, that is, if you use the `mvn:` URI scheme in a `.target` definition

[WARNING] [a7f3f9cb-793c-4f52-8e8e-b74719152388][extension>**org.eclipse.tycho:tycho-maven-plugin**:3.0.0] 

Still the message is from **tycho-maven-plugin** ;-)

Tycho+m2 both implement the mvn protocol...

**Comment by sratz** [comment]
But you're right, I mixed up some things.

This issue is about the error being visible in Eclipse:

image


The Warning in Tycho is a different issue and needs to be solved over there.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MvnProtocolHandlerService.java (hunk 1 above)
Comment (sratz) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MvnProtocolHandlerService.java:138]: JavaDoc says:
`@return  the date the resource referenced by this {@code URLConnection} was last modified, or 0 if not known.`
Also `0` is what was returned if not overriding the method at all.

Comment (laeubi) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MvnProtocolHandlerService.java:138]: Alright, then forget about that :-)

**Review by laeubi — APPROVED** [review]
A corresponding change for Tycho would be useful.

**Comment by sratz** [comment]
> A corresponding change for Tycho would be useful.

Done: https://github.com/eclipse-tycho/tycho/pull/1514

**Pull request merged**